### PR TITLE
nextfloat(x), prevfloat(x), eps(x) respect precision of BigFloat

### PR DESCRIPTION
**Pull request by dpsanders**
`nextfloat(x)`, `prevfloat(x)` and `eps(x)` now respect the precision of the `BigFloat` `x`, rather than using the global `BigFloat` precision.

This includes a rewritten generic `eps` function.

Part of #13155

### CODE REVIEW

**Comment by dpsanders** [comment]
cc @simonbyrne

File: base/float.jl [hunk 1]
```diff
@@ -703,13 +703,23 @@ end
     realmin() = realmin(Float64)
     realmax() = realmax(Float64)
 
-    eps(x::AbstractFloat) = isfinite(x) ? abs(x) >= realmin(x) ? ldexp(eps(typeof(x)),exponent(x)) : nextfloat(zero(x)) : oftype(x,NaN)
-    eps(::Type{Float16}) = $(box(Float16,unbox(UInt16,0x1400)))
-    eps(::Type{Float32}) = $(box(Float32,unbox(UInt32,0x34000000)))
-    eps(::Type{Float64}) = $(box(Float64,unbox(UInt64,0x3cb0000000000000)))
```
Comment (musm) [inline base/float.jl]: I would   leave the hard coded version for these, otherwise I think it will recompute these quantities every time.

**Comment by dpsanders** [comment]
Is this considered a feature? I.e. do I need to polish it off soonish to get into 0.6?

**Comment by tkelman** [comment]
a new keyword argument to an exported constructor looks like a feature to me

**Comment by dpsanders** [comment]
Is there any chance of getting some of these changes into 1.0?

**Comment by simonbyrne** [comment]
I'm in favour. My main changes are:

1. `eps(x::BigFloat)` should have the same precision as `x`
2. check that `eps(Float32)`/`eps(Float64)` still get reduced to constants at compile time
3. add documentation noting that these functions ignore global precision.

**Comment by dpsanders** [comment]
OK thanks - I agree.
I think that changing nextfloat and prevfloat to also respect the precision of x goes along with that?

**Comment by simonbyrne** [comment]
Yes, sorry I didn't mention that because you had already made that change!

**Comment by laborg** [comment]
@dpsanders mind to close this? As of now `nextfloat` and `prevfloat` respect the precision of `BigFloat`.

**Comment by dpsanders** [comment]
Thanks @laborg!

**Pull request closed without merging**